src/main/kotlin/org/pkl/lsp/features/HoverFeature.kt

```diff
@@ -0,0 +1,302 @@
+/**
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.lsp.features
+
+import java.util.concurrent.CompletableFuture
+import org.eclipse.lsp4j.Hover
+import org.eclipse.lsp4j.HoverParams
+import org.eclipse.lsp4j.MarkupContent
+import org.pkl.lsp.PklBaseModule
+import org.pkl.lsp.PklLSPServer
+import org.pkl.lsp.ast.*
+import org.pkl.lsp.resolvers.ResolveVisitors
+import org.pkl.lsp.resolvers.Resolvers
+import org.pkl.lsp.type.Type
+import org.pkl.lsp.type.computeResolvedImportType
+import org.pkl.lsp.type.computeThisType
+
+class HoverFeature(override val server: PklLSPServer) : Feature(server) {
+
+  fun onHover(params: HoverParams): CompletableFuture<Hover> {
+    fun run(mod: PklModule?): Hover {
+      if (mod == null) return Hover(listOf())
+      val line = params.position.line + 1
+      val col = params.position.character + 1
+      val hoverText =
+        mod.findBySpan(line, col)?.let { resolveHover(it, line, col) } ?: return Hover(listOf())
+      return Hover(MarkupContent("markdown", hoverText))
+    }
+    return server.builder().runningBuild(params.textDocument.uri).thenApply(::run)
+  }
+
+  private fun resolveHover(node: Node, line: Int, col: Int, originalNode: Node? = null): String? {
```

Comment:
FYI: the way that this works on the pkl-intellij side is that `node` is the _resolved_ value (probably by resolving the node if it is a `PsiReference`), and `originalNode` is the element that is underneath the cursor.

If you've copied over the logic in pkl-intellij, you'll want to copy the same behavior in `resolveHover`.

Reply:
Added a `resolveReference` used by the hover and go to definition.